File: 2020/Day9/Haskell/EncodingError-Part2.hs
import System.IO
import Control.Monad

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

subsets:: Int -> [a] -> [[a]]
subsets 0 _ = [[]]
subsets _ [] = []
subsets n (x : xs) = map (x :) (subsets (n - 1) xs) ++ subsets n xs

slice:: Int -> Int -> [a] -> [a]
slice from to xs = take (to - from + 1) (drop from xs)

checkSum:: [Int] -> Int -> Int
checkSum numbers index
    | numbers!!index `elem` sums = checkSum numbers (index+1)
    | otherwise = numbers!!index
    where 
        sums = map sum (subsets 2 (slice (index-25) (index-1) numbers))

findSum:: [Int] -> Int -> Int -> Int -> [Int]
findSum numbers low high target
    | sum run < target = findSum numbers low (high+1) target
    | target < sum run = findSum numbers (low+1) high target
    | otherwise = run
    where
        run = slice low high numbers

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    let numbers = map (read::String->Int) rows
    let target = checkSum numbers 25
    let run = findSum numbers 0 0 target
    print(minimum run + maximum run)
    hClose handle

File: 2020/Day09/Python/EncodingError-Part2.py
import sys
import collections
import itertools
import numpy as np

numbers = []
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    numbers.append(int(line))

target = 0

for i in range(25, len(numbers)):
    num = numbers[i]
    for a, b in itertools.combinations(numbers[i-25:i], 2):
        if a + b == num:
            break
    else:
        target = num
        break

low = high = 0

while True:
    run = numbers[low:high]
    if sum(run) < target:
        high += 1
    elif target < sum(run):
        low += 1
    else:
        break

print(min(run) + max(run))